test/test_hyperparameters.py

```diff
@@ -82,11 +83,96 @@ def test_constant(self):
+        array_2 = np.array([-2, -2])
+        array_3 = np.array([1, -2])
+        
+        wrong_shape_1 = np.array([[1]])
+        wrong_shape_2 = np.array([1, 2, 3]).reshape(1, -1)
+        wrong_shape_3 = np.array([1, 2, 3]).reshape(-1, 1)
+        
+        self.assertEqual(c1.pdf(point_1), np.array([1.0]))
+        self.assertEqual(c2.pdf(point_2), np.array([1.0]))
+        self.assertEqual(c1.pdf(point_2), np.array([0.0]))
+        self.assertEqual(c2.pdf(point_1), np.array([0.0]))
+        
+        self.assertEqual(tuple(c1.pdf(array_1)), tuple(np.array([1.0, 1.0])))
+        self.assertEqual(tuple(c2.pdf(array_2)), tuple(np.array([1.0, 1.0])))
+        self.assertEqual(tuple(c1.pdf(array_2)), tuple(np.array([0.0, 0.0])))
+        self.assertEqual(tuple(c1.pdf(array_3)), tuple(np.array([1.0, 0.0])))
+        
+        # pdf must take a numpy array
+        with self.assertRaises(TypeError):
+            c1.pdf(0.2)
+        with self.assertRaises(TypeError):
+            c1.pdf('pdf')
+        
+        # and it must be one-dimensional
+        with self.assertRaisesRegex(ValueError, "Method pdf expects a one-dimensional numpy array"):
+            c1.pdf(wrong_shape_1)
+        with self.assertRaisesRegex(ValueError, "Method pdf expects a one-dimensional numpy array"):
+            c1.pdf(wrong_shape_2)
+        with self.assertRaisesRegex(ValueError, "Method pdf expects a one-dimensional numpy array"):
+            c1.pdf(wrong_shape_3)
+            
+    def test_constant__pdf(self):
+        c1 = Constant("valuee", 1)
+        c2 = Constant("valueee", -2)
+
+        point_1 = np.array([1])
+        point_2 = np.array([-2])
+        array_1 = np.array([1, 1])
+        array_2 = np.array([-2, -2])
+        array_3 = np.array([1, -2])
+        
+        # These shapes are allowed in _pdf
+        accepted_shape_1 = np.array([[1]])
+        accepted_shape_2 = np.array([1, 2, 3]).reshape(1, -1)
+        accepted_shape_3 = np.array([3, 2, 1]).reshape(-1, 1)
+        
+        self.assertEqual(c1._pdf(point_1), np.array([1.0]))
+        self.assertEqual(c2._pdf(point_2), np.array([1.0]))
+        self.assertEqual(c1._pdf(point_2), np.array([0.0]))
+        self.assertEqual(c2._pdf(point_1), np.array([0.0]))
+        
+        # Only (N, ) numpy arrays are seamlessly converted to tuples
+        # so the __eq__ method works as intended
+        self.assertEqual(tuple(c1._pdf(array_1)), tuple(np.array([1.0, 1.0])))
+        self.assertEqual(tuple(c2._pdf(array_2)), tuple(np.array([1.0, 1.0])))
+        self.assertEqual(tuple(c1._pdf(array_2)), tuple(np.array([0.0, 0.0])))
+        self.assertEqual(tuple(c1._pdf(array_3)), tuple(np.array([1.0, 0.0])))
+
+        # pdf must take a numpy array
+        with self.assertRaises(TypeError):
+            c1._pdf(0.2)
+        with self.assertRaises(TypeError):
+            c1._pdf('pdf')
+
+        # Simply check that it runs, since _pdf does not restrict shape (only public method does)
```

Comment:
I'm not sure I follow here! Can you elaborate?

Reply:
I should have put the review comment slightly lower but here's the code for reference.

```python
# Simply check that it runs, since _pdf does not restrict shape (only public method does)
self.assertEqual(c1._pdf(accepted_shape_1)[0][0], 1.0)
self.assertEqual(c1._pdf(accepted_shape_2)[0][0], 1.0)
self.assertEqual(c1._pdf(accepted_shape_2)[0][2], 0.0)
self.assertEqual(c1._pdf(accepted_shape_3)[0][0], 0.0)
self.assertEqual(c1._pdf(accepted_shape_3)[2][0], 1.0)
```

Sure, suppose someone in the future comes along and makes some changes to `_pdf`, they run the tests and find out these lines above fail. They then might consider "oh I need to ensure these tests pass" and then code in what's required to make sure these tests above will pass.

We now have the case where `_pdf` is made to explicitly pass those tests and produce those outputs given those inputs. This means the tests have defined how `_pdf` should work. From you `#` comment above, it seems you just wanted to make sure it doesn't crash, and not actually say, this is what `_pdf` should return given a 2d input.

There's two solutions I would consider
* One is just to run `c1._pdf(accepted_shape)` without testing for output, this way you're testing that it accepts it but make no condition on what it returns.
* The other solution would be to explicitly prohibit the dimensions that can come into `_pdf` ... this sort of invalidates the checking in `_pdf` and would make me think each class just implements it's own `pdf` function and checking, leading to some slight duplication but a cleaner interface, i.e. less jumping around the code to understand a function.

I'll leave it to you to decide though, I think ConfigSpace will eventually be due for some cleaning up but it's a while away :)